source/agora/consensus/Validation.d

```diff
@@ -217,3 +219,230 @@ unittest
     // Signature verification must be error
     assert(!tx2.isValid(findOutput), format("Transaction signature is not validated %s", tx2));
 }
+
+/*******************************************************************************
+
+    Check the validity of a block.
+
+    A block is considered valid if:
+        - its height is the previous block height + 1
+        - its prev_hash is the previous block header's hash
+        - the number of transactions in the block are equal to Block.TxsInBlock
+        - the merkle root in the header matches the re-built merkle tree root
+          based on the included transactions in the block
+        - all the the transactions pass validation, which implies:
+            - signatures are authentic
+            - the inputs spend an output which must be found with the
+              findOutput() delegate
+
+    Note that checking for transactions which double-spend is the responsibility
+    of the findOutput() delegate. During validation, whenever this delegate is
+    called it should also keep track of the used UTXOs, thereby marking
+    it as a spent output. See the `findNonSpent` function in the
+    unittest for an example.
+
+    Params:
+        block = the block to check
+        prev_height = the height of the previous block which this
+                      block should point to
+        findOutput = delegate to find the referenced unspent UTXOs with
+
+    Returns:
+        true if the block is considered valid
+
+*******************************************************************************/
+
+public bool isValid (const ref Block block, in ulong prev_height,
+    in Hash prev_hash, UtxoFinder findOutput) nothrow @safe
+{
+    import std.algorithm;
+
+    // special case for the genesis block
+    if (block.header.height == 0)
+        return block == getGenesisBlock();
+
+    if (block.header.height != prev_height + 1)
+        return false;
+
+    if (block.header.prev_block != prev_hash)
+        return false;
+
+    if (block.txs.length != Block.TxsInBlock)
+        return false;
+
+    if (block.txs.any!(tx => !tx.isValid(findOutput)))
+        return false;
+
+    Hash[] merkle_tree;
+    if (block.header.merkle_root != Block.buildMerkleTree(block.txs, merkle_tree))
```

Comment:
Btw, I was thinking about this.

If the transactions in the block were already sorted lexicographically by their hash, then I could simply merge the hashes of the transactions together to create the merkle root here, and compare that to the merkle root in the header. 

Making them sorted would break a bunch of test code though.

Currently the transactions are added to the block in arbitrary order. Is this desired?